ui/src/shared/components/inlineLabels/InlineLabelsEditor.tsx

```diff
@@ -0,0 +1,231 @@
+  searchTerm: string
+  filteredLabels: Label[]
+  isPopoverVisible: boolean
+  selectedItemID: string
+  isCreatingLabel: OverlayState
+}
+
+@ErrorHandling
+class InlineLabelsEditor extends Component<Props, State> {
+  constructor(props: Props) {
+    super(props)
+
+    const initialFilteredLabels = _.differenceBy(
+      props.labels,
+      props.selectedLabels,
+      label => label.name
+    )
+
+    this.state = {
+      selectedItemID: null,
+      searchTerm: '',
+      filteredLabels: initialFilteredLabels,
+      isPopoverVisible: false,
+      isCreatingLabel: OverlayState.Hidden,
+    }
+  }
+
+  public render() {
+    const {isCreatingLabel, searchTerm} = this.state
+
+    return (
+      <>
+        <div className="inline-labels--editor">
+          <div className="inline-label--add">
+            <Button
+              color={ComponentColor.Secondary}
+              titleText="Add labels"
+              onClick={this.handleShowPopover}
+              shape={ButtonShape.Square}
+              icon={IconFont.Plus}
+            />
+          </div>
+          {this.popover}
+          {this.noLabelsIndicator}
+        </div>
+        <CreateLabelOverlay
+          isVisible={isCreatingLabel === OverlayState.Visible}
+          onDismiss={this.handleStopCreatingLabel}
+          overrideDefaultName={searchTerm}
+          onCreateLabel={this.handleCreateLabel}
+          onNameValidation={this.handleEnsureUniqueLabelName}
+        />
+      </>
+    )
+  }
+
+  private get popover(): JSX.Element {
+    const {labels, selectedLabels} = this.props
+    const {searchTerm, isPopoverVisible, selectedItemID} = this.state
+
+    const labelsUsed = labels.length === selectedLabels.length
+
+    if (isPopoverVisible) {
+      return (
+        <InlineLabelPopover
+          searchTerm={searchTerm}
+          selectedItemID={selectedItemID}
+          allLabelsUsed={labelsUsed}
+          onDismiss={this.handleDismissPopover}
+          onStartCreatingLabel={this.handleStartCreatingLabel}
+          onInputChange={this.handleInputChange}
+          filteredLabels={this.availableLabels}
+          onAddLabel={this.handleAddLabel}
+          onUpdateSelectedItem={this.handleUpdateSelectedItem}
+        />
+      )
+    }
+  }
+
+  private get noLabelsIndicator(): JSX.Element {
+    const {selectedLabels} = this.props
+
+    if (selectedLabels.length) {
+      return
+    }
+
+    return (
+      <div
+        className="label label--xs label--colorless"
+        onClick={this.handleShowPopover}
+      >
+        Add a label
+      </div>
+    )
+  }
+
+  private handleAddLabel = (labelID: string): void => {
+    const {onAddLabel, labels} = this.props
+
+    const label = labels.find(label => label.id === labelID)
+
+    onAddLabel(label)
+  }
+
+  private handleUpdateSelectedItem = (selectedItemID: string): void => {
+    this.setState({selectedItemID})
+  }
+
+  private handleShowPopover = () => {
+    const {availableLabels} = this
```

Comment:
nit: is this destructuring necessary since `this.availableLabels` is being called again in line 162